File: backend/src/utils/textMeasure.ts
import opentype from 'opentype.js';
import path from 'path';

// Font weight → file mapping
const FONT_FILES: Record<string, string> = {
  '400': 'Kanit-Regular.ttf',
  '700': 'Kanit-Bold.ttf',
  '900': 'Kanit-Black.ttf',
};

// Normalize weight aliases to canonical values
const WEIGHT_ALIASES: Record<string, string> = {
  normal: '400',
  regular: '400',
  bold: '700',
  black: '900',
};

// Lazy-loaded font cache (singleton per weight)
const fontCache = new Map<string, opentype.Font>();

function getFont(fontWeight: string = '400'): opentype.Font {
  // Normalize weight
  const normalized = WEIGHT_ALIASES[fontWeight.toLowerCase()] ?? fontWeight;
  const weight = FONT_FILES[normalized] ? normalized : '400';

  const cached = fontCache.get(weight);
  if (cached) return cached;

  const filename = FONT_FILES[weight];
  const fontPath = path.resolve(__dirname, '../../assets/fonts/', filename);
  const font = opentype.loadSync(fontPath);
  fontCache.set(weight, font);
  return font;
}

export interface TextMetrics {
  width: number;
  height: number;
  ascent: number;
  descent: number;
}

export function measureText(
  text: string,
  fontSize: number,
  fontWeight?: string
): TextMetrics {
  const font = getFont(fontWeight);
  const width = font.getAdvanceWidth(text, fontSize);
  const ascent = (font.ascender / font.unitsPerEm) * fontSize;
  const descent = Math.abs(font.descender / font.unitsPerEm) * fontSize;
  const height = ascent + descent;
  return { width, height, ascent, descent };
}

export interface WrapResult {
  lines: string[];
  lineHeight: number;
  totalHeight: number;
}

export function wrapText(
  text: string,
  maxWidth: number,
  fontSize: number,
  fontWeight?: string
): WrapResult {
  const lineHeight = fontSize * 1.35;
  const lines: string[] = [];

  // Split by explicit newlines first
  const paragraphs = text.split('\n');

  for (const paragraph of paragraphs) {
    if (paragraph === '') {
      lines.push('');
      continue;
    }

    const measured = measureText(paragraph, fontSize, fontWeight);
    if (measured.width <= maxWidth) {
      lines.push(paragraph);
      continue;
    }

    // Word-based splitting
    const words = paragraph.split(/\s+/);
    let currentLine = '';

    for (const word of words) {
      const candidate = currentLine ? `${currentLine} ${word}` : word;
      const candidateWidth = measureText(candidate, fontSize, fontWeight).width;

      if (candidateWidth <= maxWidth) {
        currentLine = candidate;
      } else {
        // Push current line if non-empty
        if (currentLine) {
          lines.push(currentLine);
        }

        // Check if the single word fits
        const wordWidth = measureText(word, fontSize, fontWeight).width;
        if (wordWidth <= maxWidth) {
          currentLine = word;
        } else {
          // Force-break at character level
          let charLine = '';
          for (const char of word) {
            const charCandidate = charLine + char;
            if (measureText(charCandidate, fontSize, fontWeight).width <= maxWidth) {
              charLine = charCandidate;
            } else {
              if (charLine) lines.push(charLine);
              charLine = char;
            }
          }
          currentLine = charLine;
        }
      }
    }

    if (currentLine) {
      lines.push(currentLine);
    }
  }

  return {
    lines,
    lineHeight,
    totalHeight: lines.length * lineHeight,
  };
}

export interface AutoFitResult {
  fontSize: number;
  width: number;
}

export function autoFitFontSize(
  text: string,
  maxWidth: number,
  maxFontSize: number,
  minFontSize: number,
  fontWeight?: string
): AutoFitResult {
  let lo = minFontSize;
  let hi = maxFontSize;
  let bestSize = minFontSize;
  let bestWidth = measureText(text, minFontSize, fontWeight).width;

  // Binary search for largest fontSize where width <= maxWidth
  while (hi - lo >= 1) {
    const mid = Math.floor((lo + hi) / 2);
    const w = measureText(text, mid, fontWeight).width;

    if (w <= maxWidth) {
      bestSize = mid;
      bestWidth = w;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }

  return { fontSize: bestSize, width: bestWidth };
}

File: backend-python/app/utils/text_measure.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from fontTools.ttLib import TTFont

FONT_DIR = Path(__file__).parent.parent.parent / "assets" / "fonts"

FONT_FILES = {"400": "Kanit-Regular.ttf", "700": "Kanit-Bold.ttf", "900": "Kanit-Black.ttf"}
WEIGHT_ALIASES = {"normal": "400", "regular": "400", "bold": "700", "black": "900"}

_font_cache: dict[str, TTFont] = {}


def _get_font(weight: str = "400") -> TTFont:
    normalized = WEIGHT_ALIASES.get(weight.lower(), weight)
    w = normalized if normalized in FONT_FILES else "400"
    if w not in _font_cache:
        _font_cache[w] = TTFont(str(FONT_DIR / FONT_FILES[w]))
    return _font_cache[w]


def _glyph_width(font: TTFont, text: str, font_size: float) -> float:
    cmap = font.getBestCmap()
    hmtx = font["hmtx"]
    units_per_em = font["head"].unitsPerEm
    total = 0
    for ch in text:
        gid = cmap.get(ord(ch))
        if gid:
            total += hmtx[gid][0]
        else:
            total += units_per_em * 0.5
    return total * font_size / units_per_em


@dataclass
class TextMetrics:
    width: float
    height: float
    ascent: float
    descent: float


def measure_text(text: str, font_size: float, font_weight: str = "400") -> TextMetrics:
    font = _get_font(font_weight)
    width = _glyph_width(font, text, font_size)
    units_per_em = font["head"].unitsPerEm
    os2 = font["OS/2"]
    ascent = os2.sTypoAscender * font_size / units_per_em
    descent = abs(os2.sTypoDescender) * font_size / units_per_em
    return TextMetrics(width=width, height=ascent + descent, ascent=ascent, descent=descent)


@dataclass
class WrapResult:
    lines: list[str]
    line_height: float
    total_height: float


def wrap_text(text: str, max_width: float, font_size: float, font_weight: str = "400") -> WrapResult:
    line_height = font_size * 1.35
    lines: list[str] = []

    for paragraph in text.split("\n"):
        if not paragraph:
            lines.append("")
            continue

        if measure_text(paragraph, font_size, font_weight).width <= max_width:
            lines.append(paragraph)
            continue

        words = paragraph.split()
        current_line = ""
        for word in words:
            candidate = f"{current_line} {word}" if current_line else word
            if measure_text(candidate, font_size, font_weight).width <= max_width:
                current_line = candidate
            else:
                if current_line:
                    lines.append(current_line)
                if measure_text(word, font_size, font_weight).width <= max_width:
                    current_line = word
                else:
                    char_line = ""
                    for ch in word:
                        if measure_text(char_line + ch, font_size, font_weight).width <= max_width:
                            char_line += ch
                        else:
                            if char_line:
                                lines.append(char_line)
                            char_line = ch
                    current_line = char_line
        if current_line:
            lines.append(current_line)

    return WrapResult(lines=lines, line_height=line_height, total_height=len(lines) * line_height)


@dataclass
class AutoFitResult:
    font_size: int
    width: float


def auto_fit_font_size(
    text: str, max_width: float, max_font_size: int, min_font_size: int, font_weight: str = "400"
) -> AutoFitResult:
    lo, hi = min_font_size, max_font_size
    best_size = min_font_size
    best_width = measure_text(text, min_font_size, font_weight).width

    while hi - lo >= 1:
        mid = (lo + hi) // 2
        w = measure_text(text, mid, font_weight).width
        if w <= max_width:
            best_size = mid
            best_width = w
            lo = mid + 1
        else:
            hi = mid - 1

    return AutoFitResult(font_size=best_size, width=best_width)
